api/src/main/java/org/apache/iceberg/io/LocationProvider.java

```diff
@@ -29,6 +29,12 @@
  * Implementations must be {@link Serializable} because instances will be serialized to tasks.
  */
 public interface LocationProvider extends Serializable {
+
+  /**
+   * Return a fully-qualified data location.
+   */
+  String dataLocation();
```

Comment:
What about providers that don't have a reliable location? Do they return null?

Reply:
Instead of requiring a data location, why not use `table.location()`?

Reply:
My assumption was that there would always be at least some base reliable data location. Is that assumption false?

Technically, data location can be arbitrary today and might be outside `table.location()`. That would be odd but possible.

Reply:
I think we should clean `table.location()` by default and allow the user to configure a different path. That covers all reasonable cases.

It should be safe to clean `table.location()` because we need to protect all metadata file paths individually anyway, instead of using a prefix. So we may as well clean the whole table location.

Reply:
Yeah, that's what I was thinking as well. If we spend a bit more effort to also collect all metadata files, we can clean the table location completely.

Reply:
Done.